### PR TITLE
Speed-up serialization of TS components

### PR DESCRIPTION
**Pull request by michaelrambeau**
## Goal

Speed up the serialization of TypeScript component because it takes about 0.5 seconds by component, which is ridiculous for design systems with a lot of components (more than 50 components in Porsche and DXC design systems)

It should make the experimental mode faster as the serialization runs every time a component is saved.

## Strategy

What took time was the creation of the TypeScript program (`ts.createProgram`), that was repeated for every component.
The serialization itself only takes a few ms.

In this PR, the `program` is created only once, passing the paths of all TS components.
To achieve this, I introduced a class `Serializer`whose constructor expects the array of all the components to serialize later.

```ts
// Before
getComponentMetadata(component)

// After: a serializer object handles the context for all components
const serializer = new MergeComponentSerializer(components);
await serializer.init();

serializer.serialize(component)
```

## How to check

Run the experimental mode in debug mode to check the response times

```
DEBUG=uxpin* yarn uxpin --skip-browser
```

Save a file and check the bundle is ready much faster than before.

Ensure it works as before for TS and JS components.

### Before

Check the numbers in red and note that the serialization of each component takes between 500ms and 800ms!

image

### After

Notice the creation of the TS program at the beginning

image

### CODE REVIEW

File: packages/uxpin-merge-cli/src/steps/serialization/serializer.ts
Comment (Evomatic) [inline packages/uxpin-merge-cli/src/steps/serialization/serializer.ts:1]: So is the reasons for serialisation being slow before, due to computation being called for each component?

**Review by Evomatic — APPROVED** [review]
LGTM 🧑‍💻 

I tested with DXC Halstack and MUI-Merge. Both appear to be loading into experimental faster, as well as better reload when adjusting DS code!

File: packages/uxpin-merge-cli/src/steps/serialization/serializer.ts
Comment (michaelrambeau) [reply packages/uxpin-merge-cli/src/steps/serialization/serializer.ts:1]: It's the call to `ts.createProgram`, repeated for all components that used to take a lot of time.
Check the second screenshot: now it takes about 1 second but it's called only once, at the start.

File: packages/uxpin-merge-cli/src/steps/serialization/component/implementation/typescript/component/findComponentFile.ts [hunk 1]
```diff
@@ -2,7 +2,9 @@ import * as ts from 'typescript';
 
 export function findComponentFile(program: ts.Program, path: string): ts.SourceFile | undefined {
   for (const sourceFile of program.getSourceFiles()) {
-    if (sourceFile.fileName === path) {
+    // for an unknown reason `sourceFile.fileName` can be either a relative path (`src/xxx.tsx`)
+    // or an absolute file path (`Users/yyy/src/xxx.tsx`)
+    if (sourceFile.fileName.endsWith(path)) {
```
[the remark this answers is not in the document]
Comment (michaelrambeau) [reply packages/uxpin-merge-cli/src/steps/serialization/component/implementation/typescript/component/findComponentFile.ts:7]: Like I said in the comment, I don't know the reason why for some components I get an absolute path when looping through `program.getSourceFiles()`, for DXC.

E.g. Accordion.tsx, Footer.tsx, Header.tsx...

My assumption is these components are imported by other components, they are discovered by the compiler when checking other components, before being discovered by themselves.
For example, the compiler finds `AccordionGroup` before `Accordion` and as a result `Accordion` get registered with a full path. But it's jus an assumption.

Here is the list I get:

```
src/utilities/advanced-theme-generator/AdvancedThemeGenerator.tsx
src/components/table/tbody/TableBody.tsx
src/components/table/td/TableDataCell.tsx
src/components/table/tfoot/TableFoot.tsx
src/components/table/th/TableHeader.tsx
src/components/table/thead/TableHead.tsx
src/components/table/tr/TableRow.tsx
src/components/accordion/Accordion.tsx
src/components/accordion-group/AccordionGroup.tsx
/Users/***/uxpin-merge/halstack-uxpin-merge/src/components/accordion-group/accordion/Accordion.tsx
src/components/alert/Alert.tsx
src/components/application-layout/ApplicationLayout.tsx
/Users/***/uxpin-merge/halstack-uxpin-merge/src/components/application-layout/footer/Footer.tsx
/Users/***/uxpin-merge/halstack-uxpin-merge/src/components/application-layout/header/Header.tsx
/Users/***/uxpin-merge/halstack-uxpin-merge/src/components/application-layout/header/dropdown/Dropdown.tsx
/Users/***/uxpin-merge/halstack-uxpin-merge/src/components/application-layout/main/Main.tsx
/Users/***/uxpin-merge/halstack-uxpin-merge/src/components/application-layout/sidenav/Sidenav.tsx
/Users/***/uxpin-merge/halstack-uxpin-merge/src/components/application-layout/sidenav/sidenav-group/Group.tsx
/Users/***/uxpin-merge/halstack-uxpin-merge/src/components/application-layout/sidenav/sidenav-link/Link.tsx
/Users/***/uxpin-merge/halstack-uxpin-merge/src/components/application-layout/sidenav/sidenav-section/Section.tsx
/Users/***/uxpin-merge/halstack-uxpin-merge/src/components/application-layout/sidenav/sidenav-title/Title.tsx
src/components/bleed/Bleed.tsx
src/components/box/Box.tsx
src/components/bulleted-list/BulletedList.tsx
/Users/***/uxpin-merge/halstack-uxpin-merge/src/components/bulleted-list/item/Item.tsx
src/components/button/Button.tsx
src/components/card/Card.tsx
src/components/checkbox/Checkbox.tsx
src/components/chip/Chip.tsx
src/components/date-input/DateInput.tsx
src/components/dialog/Dialog.tsx
src/components/dropdown/Dropdown.tsx
src/components/file-input/FileInput.tsx
src/components/flex/Flex.tsx
src/components/grid/Grid.tsx
/Users/***/uxpin-merge/halstack-uxpin-merge/src/components/grid/grid-item/GridItem.tsx
src/components/heading/Heading.tsx
src/components/inset/Inset.tsx
src/components/link/Link.tsx
src/components/nav-tabs/NavTabs.tsx
/Users/***/uxpin-merge/halstack-uxpin-merge/src/components/nav-tabs/tab/Tab.tsx
src/components/number-input/NumberInput.tsx
src/components/paginator/Paginator.tsx
src/components/paragraph/Paragraph.tsx
src/components/password-input/PasswordInput.tsx
src/components/progress-bar/ProgressBar.tsx
src/components/quick-nav/QuickNav.tsx
src/components/radio-group/RadioGroup.tsx
src/components/resultset-table/ResultsetTable.tsx
src/components/select/Select.tsx
src/components/slider/Slider.tsx
src/components/spinner/Spinner.tsx
src/components/switch/Switch.tsx
/Users/***/uxpin-merge/halstack-uxpin-merge/src/components/table/Table.tsx
src/components/tabs/Tabs.tsx
src/components/tag/Tag.tsx
src/components/textarea/Textarea.tsx
src/components/text-input/TextInput.tsx
src/components/toggle-group/ToggleGroup.tsx
src/components/typography/Typography.tsx
src/components/wizard/Wizard.tsx
```

**Review by Evomatic — APPROVED** [review]

**Pull request merged**